hedera-node/hedera-app/src/main/java/com/hedera/node/app/records/files/StreamFileProducerConcurrent.java

```diff
@@ -0,0 +1,318 @@
+
+    /** The hapi version of the node */
+    private final SemanticVersion hapiVersion;
+    /** The executor service to use for background tasks */
+    private final ExecutorService executorService;
+    /** Future for running hash results of last running hash updates task */
+    private CompletableFuture<Bytes> lastRecordHashingResult = null;
+    /** Future for running hash results of previous lastRecordHashingResult */
+    private CompletableFuture<Bytes> lastRecordHashingResultNMinus1 = null;
+    /** Future for running hash results of previous, previous lastRecordHashingResult */
+    private CompletableFuture<Bytes> lastRecordHashingResultNMinus2 = null;
+    /** Future for running hash results of previous, previous, previous lastRecordHashingResult */
+    private CompletableFuture<Bytes> lastRecordHashingResultNMinus3 = null;
+    /** Future for the current record file writer, this will always get you the record file writer for teh current file being written */
+    private CompletableFuture<RecordFileWriter> currentRecordFileWriter = null;
+    /** Future to get list of current sidecar file writers, starts with completed empty list to save handing null at start */
+    private CompletableFuture<List<SidecarFileWriter>> sidecarHandlingFuture = CompletableFuture.completedFuture(null);
+
+    /**
+     * Construct StreamFileProducerConcurrent
+     *
+     * @param configProvider the configuration to read from
+     * @param nodeInfo the current node information
+     * @param signer the signer to use for signing in signature files
+     * @param fileSystem the file system to use, needed for testing to be able to use a non-standard file
+     *                   system. If null default is used.
+     * @param executorService The executor service to use for background threads
+     */
+    public StreamFileProducerConcurrent(
+            @NonNull final ConfigProvider configProvider,
+            @NonNull final NodeInfo nodeInfo,
+            @NonNull final Signer signer,
+            @Nullable final FileSystem fileSystem,
+            @NonNull final ExecutorService executorService) {
+        super(configProvider, nodeInfo, signer, fileSystem);
+        this.executorService = executorService;
+        this.hapiVersion = nodeInfo.hapiVersion();
+    }
+
+    /**
+     * Set the current running hash of record stream items. This is called only once to initialize the running hash on startup.
+     * This is called on handle transaction thread.
+     *
+     * @param recordStreamItemRunningHash The new running hash, all future running hashes produced will flow from this
+     */
+    @Override
+    public void setRunningHash(Bytes recordStreamItemRunningHash) {
+        lastRecordHashingResult = CompletableFuture.completedFuture(recordStreamItemRunningHash);
+        lastRecordHashingResultNMinus1 = null;
+        lastRecordHashingResultNMinus2 = null;
+        lastRecordHashingResultNMinus3 = null;
+    }
```

Comment:
Maybe we should rename this method to "initializeRunningHash" or "setInitialRunningHash" or something, since it is only called from BlockRecordManagerImpl in a very specific place during initialization.

I'm also a little confused by it. I am trying to determine if it should be `@Nullable` or `@NonNull`. The documentation suggests it is loaded at startup, I'm assuming from a saved state (i.e. we had the hash saved to state, and so we need to load it from state in here, so we can start our computation). But then shouldn't there also be N-1, N-2, and N-3 hashes also saved in state that we would load here?